allennlp/training/supervised_trainer.py

```diff
@@ -0,0 +1,816 @@
+            metrics["training_epochs"] = epochs_trained
+            metrics["epoch"] = epoch
+
+            for key, value in train_metrics.items():
+                metrics["training_" + key] = value
+            for key, value in val_metrics.items():
+                metrics["validation_" + key] = value
+
+            if is_best_so_far:
+                # Update all the best_ metrics.
+                # (Otherwise they just stay the same as they were.)
+                metrics['best_epoch'] = epoch
+                for key, value in val_metrics.items():
+                    metrics["best_validation_" + key] = value
+
+            if self._serialization_dir:
+                dump_metrics(os.path.join(self._serialization_dir, f'metrics_epoch_{epoch}.json'), metrics)
+
+            if self._learning_rate_scheduler:
+                # The LRScheduler API is agnostic to whether your schedule requires a validation metric -
+                # if it doesn't, the validation metric passed here is ignored.
+                self._learning_rate_scheduler.step(this_epoch_val_metric, epoch)
+
+            self._save_checkpoint(epoch, validation_metric_per_epoch, is_best=is_best_so_far)
+
+            epoch_elapsed_time = time.time() - epoch_start_time
+            logger.info("Epoch duration: %s", time.strftime("%H:%M:%S", time.gmtime(epoch_elapsed_time)))
+
+            if epoch < self._num_epochs - 1:
+                training_elapsed_time = time.time() - training_start_time
+                estimated_time_remaining = training_elapsed_time * \
+                    ((self._num_epochs - epoch_counter) / float(epoch - epoch_counter + 1) - 1)
+                formatted_time = str(datetime.timedelta(seconds=int(estimated_time_remaining)))
+                logger.info("Estimated training time remaining: %s", formatted_time)
+
+            epochs_trained += 1
+
+        return metrics
+
+    def _is_best_so_far(self,
+                        this_epoch_val_metric: float,
+                        validation_metric_per_epoch: List[float]):
+        if not validation_metric_per_epoch:
+            return True
+        elif self._validation_metric_decreases:
+            return this_epoch_val_metric < min(validation_metric_per_epoch)
+        else:
+            return this_epoch_val_metric > max(validation_metric_per_epoch)
+
+    def _description_from_metrics(self, metrics: Dict[str, float]) -> str:
+        if (not self._warned_tqdm_ignores_underscores and
+                    any(metric_name.startswith("_") for metric_name in metrics)):
+            logger.warning("Metrics with names beginning with \"_\" will "
+                           "not be logged to the tqdm progress bar.")
+            self._warned_tqdm_ignores_underscores = True
+        return ', '.join(["%s: %.4f" % (name, value) for name, value in
+                          metrics.items() if not name.startswith("_")]) + " ||"
+
+    def _save_checkpoint(self,
```

Comment:
Say I want to make a `MultiTaskTrainer`.  What's the right way to get this function included?  Should it be in the base `Trainer` class?  Or should it be factored out into a utility function that this internal method calls?  Or should I inherit from `SingleTaskTrainer`?  It seems like there are lots of methods in here where this question applies.